### PR TITLE
Implement QScript joins.

### LINKED ISSUE #1560: compile a join with an explicit join condition to QScript
```
    "convert basic join with explicit join condition" in {
      //"select foo.name, bar.address from foo join bar on foo.id = bar.foo_id",

      val lp = LP.Let('__tmp0, lpRead("/foo"),
        LP.Let('__tmp1, lpRead("/bar"),
          LP.Let('__tmp2,
            StdLib.set.InnerJoin(LP.Free('__tmp0), LP.Free('__tmp1),
              relations.Eq(
                structural.ObjectProject(LP.Free('__tmp0), LP.Constant(Data.Str("id"))).embed,
                structural.ObjectProject(LP.Free('__tmp1), LP.Constant(Data.Str("foo_id"))).embed).embed).embed,
            makeObj(
              "name" ->
                structural.ObjectProject(
                  structural.ObjectProject(LP.Free('__tmp2), LP.Constant(Data.Str("left"))).embed,
                  LP.Constant(Data.Str("name"))),
              "address" ->
                structural.ObjectProject(
                  structural.ObjectProject(LP.Free('__tmp2), LP.Constant(Data.Str("right"))).embed,
                  LP.Constant(Data.Str("address")))))))
      convert(None, lp) must
      equal(chain(
        RootR,
        QC.inj(Map((), ProjectFieldR(HoleF, StrLit("foo"))))).some)
    }
```

This test is `pendingUntilFixed`.

### LINKED ISSUE #1581: Query with join creates invalid QScript
Copy-paste conversation with @sellout 

for
`select * from /rabbit/hadoop/test4.txt as left join /rabbit/hadoop/test2.txt as right on left.country = right.country`
I get 0 results

sellout [1:06]  
I simplified it down to this:

``` Subset(
  Unreferenced,
  EquiJoin(
    Unreferenced,
      Map(ShiftedRead(/rabbit/hadoop/test4.txt, ExcludeId),
        Guard(SrcHole, Obj(Map(),Some(Top)), SrcHole, Undefined())),
      Map(ShiftedRead(/rabbit/hadoop/test2.txt, ExcludeId),
        Guard(SrcHole, Obj(Map(),Some(Top)), SrcHole, Undefined())),
      MakeArray(SrcHole),
      MakeArray(SrcHole),
      Inner,
      Guard(RightSide, Obj(Map(),Some(Top)), Guard(LeftSide, Obj(Map(),Some(Top)), ConcatMaps(LeftSide, RightSide), Undefined()), Undefined())),
  Take,
  Map(Unreferenced,Constant(Int(11))))
```

[1:06]  
Which looks right except that the keys are `MakeArray(SrcHole)` when they should be `MakeArray(ProjectField(SrcHole, StrLit("country")))`.

[1:07]  
I.e., it’s only going to return a result when both sides are equal.

Long version of generated QS

```
QScript:
List
├─ Unreferenced
╰─ Subset((),ThetaJoin(Unreferenced,LeftShift(Read(/rabbit/hadoop/test4.txt),ZipMapKeys(SrcHole),ConcatArrays(MakeArray(ConcatArrays(ConcatArrays(ConcatArrays(MakeArray(MakeMap(Constant(Str(m)), ProjectIndex(RightSide, Constant(Int(0))))), Constant(Arr(List(Fix(Coproduct(-\/(Map(List((Fix(Coproduct(\/-(Str(f)))),Fix(Coproduct(\/-(Str(test4.txt)))))))))))))), Constant(Arr(List(Fix(Coproduct(-\/(Map(List((Fix(Coproduct(\/-(Str(f)))),Fix(Coproduct(\/-(Str(hadoop)))))))))))))), Constant(Arr(List(Fix(Coproduct(-\/(Map(List((Fix(Coproduct(\/-(Str(f)))),Fix(Coproduct(\/-(Str(rabbit))))))))))))))), MakeArray(Guard(ProjectIndex(RightSide, Constant(Int(1))), Obj(Map(),Some(Top)), ProjectIndex(RightSide, Constant(Int(1))), Undefined())))),LeftShift(Read(/rabbit/hadoop/test2.txt),ZipMapKeys(SrcHole),ConcatArrays(MakeArray(ConcatArrays(ConcatArrays(ConcatArrays(MakeArray(MakeMap(Constant(Str(m)), ProjectIndex(RightSide, Constant(Int(0))))), Constant(Arr(List(Fix(Coproduct(-\/(Map(List((Fix(Coproduct(\/-(Str(f)))),Fix(Coproduct(\/-(Str(test2.txt)))))))))))))), Constant(Arr(List(Fix(Coproduct(-\/(Map(List((Fix(Coproduct(\/-(Str(f)))),Fix(Coproduct(\/-(Str(hadoop)))))))))))))), Constant(Arr(List(Fix(Coproduct(-\/(Map(List((Fix(Coproduct(\/-(Str(f)))),Fix(Coproduct(\/-(Str(rabbit))))))))))))))), MakeArray(Guard(ProjectIndex(RightSide, Constant(Int(1))), Obj(Map(),Some(Top)), ProjectIndex(RightSide, Constant(Int(1))), Undefined())))),Eq(LeftSide, RightSide),Inner,Guard(ProjectIndex(RightSide, Constant(Int(1))), Obj(Map(),Some(Top)), Guard(ProjectIndex(LeftSide, Constant(Int(1))), Obj(Map(),Some(Top)), ConcatMaps(ProjectIndex(LeftSide, Constant(Int(1))), ProjectIndex(RightSide, Constant(Int(1)))), Undefined()), Undefined())),Take,Map(Unreferenced,Constant(Int(11))))
   ╰─ ExternallyManaged(Extern)

QScript (Spark):
Subset((),Map(EquiJoin(Unreferenced,Map(ShiftedRead(/rabbit/hadoop/test4.txt, IncludeId),ConcatArrays(MakeArray(ConcatArrays(ConcatArrays(ConcatArrays(MakeArray(MakeMap(Constant(Str(m)), ProjectIndex(SrcHole, Constant(Int(0))))), Constant(Arr(List(Fix(Coproduct(-\/(Map(List((Fix(Coproduct(\/-(Str(f)))),Fix(Coproduct(\/-(Str(test4.txt)))))))))))))), Constant(Arr(List(Fix(Coproduct(-\/(Map(List((Fix(Coproduct(\/-(Str(f)))),Fix(Coproduct(\/-(Str(hadoop)))))))))))))), Constant(Arr(List(Fix(Coproduct(-\/(Map(List((Fix(Coproduct(\/-(Str(f)))),Fix(Coproduct(\/-(Str(rabbit))))))))))))))), MakeArray(Guard(ProjectIndex(SrcHole, Constant(Int(1))), Obj(Map(),Some(Top)), ProjectIndex(SrcHole, Constant(Int(1))), Undefined())))),Map(ShiftedRead(/rabbit/hadoop/test2.txt, IncludeId),ConcatArrays(MakeArray(ConcatArrays(ConcatArrays(ConcatArrays(MakeArray(MakeMap(Constant(Str(m)), ProjectIndex(SrcHole, Constant(Int(0))))), Constant(Arr(List(Fix(Coproduct(-\/(Map(List((Fix(Coproduct(\/-(Str(f)))),Fix(Coproduct(\/-(Str(test2.txt)))))))))))))), Constant(Arr(List(Fix(Coproduct(-\/(Map(List((Fix(Coproduct(\/-(Str(f)))),Fix(Coproduct(\/-(Str(hadoop)))))))))))))), Constant(Arr(List(Fix(Coproduct(-\/(Map(List((Fix(Coproduct(\/-(Str(f)))),Fix(Coproduct(\/-(Str(rabbit))))))))))))))), MakeArray(Guard(ProjectIndex(SrcHole, Constant(Int(1))), Obj(Map(),Some(Top)), ProjectIndex(SrcHole, Constant(Int(1))), Undefined())))),MakeArray(SrcHole),MakeArray(SrcHole),Inner,ConcatArrays(MakeArray(LeftSide), MakeArray(RightSide))),Guard(ProjectIndex(ProjectIndex(SrcHole, Constant(Int(1))), Constant(Int(1))), Obj(Map(),Some(Top)), Guard(ProjectIndex(ProjectIndex(SrcHole, Constant(Int(0))), Constant(Int(1))), Obj(Map(),Some(Top)), ConcatMaps(ProjectIndex(ProjectIndex(SrcHole, Constant(Int(0))), Constant(Int(1))), ProjectIndex(ProjectIndex(SrcHole, Constant(Int(1))), Constant(Int(1)))), Undefined()), Undefined())),Take,Map(Unreferenced,Constant(Int(11))))
╰─ Unreferenced
```

### PR DESCRIPTION
**Pull request by alissapajer**
Fixes #1581 
Fixes #1560 

Implements basic joins in QScript.

### CODE REVIEW

File: connector/src/main/scala/quasar/qscript/Rewrite.scala [hunk 1]
```diff
@@ -69,6 +69,29 @@ class Rewrite[T[_[_]]: BirecursiveT: EqualT] extends TTypes[T] {
     case x                                 => FtoG(QC.inj(x))
   }
 
+  // TODO unify this function with `unifySimpleBranches`
+  // including all the cases
```
Comment (alissapajer) [inline connector/src/main/scala/quasar/qscript/Rewrite.scala]: The case implemented below is the case we hit in the basic joins I'm testing here, but really we should just generalize `unifySimpleBranches` to be able to return what we need in either case.

File: it/src/main/resources/tests/sameFieldName.test [hunk 2]
```diff
@@ -2,7 +2,6 @@
     "name": "select over fields with same name",
     "backends": {
         "couchbase":         "skip",
-        "marklogic":         "skip",
```
Comment (alissapajer) [inline it/src/main/resources/tests/sameFieldName.test]: yay :)

Comment (sellout) [inline it/src/main/resources/tests/sameFieldName.test]: So much yay.

File: connector/src/main/scala/quasar/qscript/Transform.scala [hunk 3]
```diff
@@ -381,38 +432,103 @@ class Transform
...
+
+        val Ann(leftBuckets, leftValue) = values(0).ann
+        val Ann(rightBuckets, rightValue) = values(1).ann
+
+        // NB: #1556 This is a magic structure. Improve LogicalPlan to not imply this structure.
+        val combine: JoinFunc = Free.roll(ConcatMaps(
+          Free.roll(MakeMap(StrLit[T, JoinSide](JoinDir.Left.name), leftValue.as(LeftSide))),
+          Free.roll(MakeMap(StrLit[T, JoinSide](JoinDir.Right.name), rightValue.as(RightSide)))))
+
+        val pair: PlannerError \/ (SrcMerge[FreeQS, FreeMap], SrcMerge[FreeQS, FreeMap]) =
```
Comment (sellout) [inline connector/src/main/scala/quasar/qscript/Transform.scala]: Probably add a FIXME here about the guessing which side is which (which should hopefully be fixed along with #1556).

Comment (alissapajer) [inline connector/src/main/scala/quasar/qscript/Transform.scala]: Good idea.

**Comment by alissapajer** [comment]
@sellout change pushed

**Pull request merged**